### PR TITLE
Update predict production   PM2.5 fallback for site forecast inputs

### PR DESCRIPTION
**Pull request by wabinyai**
# 🚀 Pull Request

### What does this PR do?

- Updated the daily site forecast BigQuery query to use `COALESCE(pm2_5_calibrated_value, pm2_5)` for `pm25_mean`, `pm25_min`, and `pm25_max`.
- Updated the hourly site forecast BigQuery query to use `COALESCE(pm2_5_calibrated_value, pm2_5)` for `pm25_mean`.
- Updated the non-null filters so rows are included when either calibrated PM2.5 or raw PM2.5 is available.
- Added tests to confirm both hourly and daily site forecast queries use the PM2.5 fallback.

### Why is this change needed?

The site-level hourly and daily forecasting DAGs currently rely only on `pm2_5_calibrated_value` from BigQuery. Some site records have `pm2_5_calibrated_value` as `NULL` while still having valid raw `pm2_5` values. Those rows were being excluded from forecast input data, which could reduce available history per `site_id` and cause sites to miss forecast generation.
---

## 🔗 Related Issues
- [ ] Closes #
- [ ] Fixes #
- [ ] Related to #

---

## 🔄 Type of Change
- [ ] 🐛 Bug fix
- [ ] ✨ New feature
- [ ] 🔧 Enhancement/improvement
- [ ] 📚 Documentation update
- [ ] ♻️ Refactor
- [ ] 🗑️ Removal/deprecation

---

## 🏗️ Affected Services
**Microservices changed:**

---

## 🧪 Testing
- [ ] Unit tests added/updated
- [ ] Manual testing completed
- [ ] All existing tests pass

**Test summary:**

---

## 💥 Breaking Changes
- [ ] **No breaking changes**
- [ ] **Has breaking changes** (describe below)

---

## 📝 Additional Notes

---

## ✅ Checklist
- [ ] Code follows project style guidelines
- [ ] Self-review completed
- [ ] Documentation updated (if needed)
- [ ] Ready for review

## Summary by CodeRabbit

* **Bug Fixes**
  * PM2.5 aggregations for forecasting now prefer calibrated values and fall back to raw; site name and site location selection logic normalized to prefer authoritative coordinates.

* **New Features**
  * Hourly forecast MET gap-filling: missing temperature and wind values are backfilled using the nearest available MET observation by time/location.

* **Tests**
  * Expanded tests covering PM2.5 aggregation behavior, location selection, and MET gap-filling for hourly and daily forecasts.

* **Documentation**
  * Added comments documenting calibrated-vs-raw PM2.5 fallback behavior.

### CODE REVIEW

File: src/workflows/airqo_etl_utils/tests/test_site_hourly_forecast.py [hunk 1]
```diff
@@ -561,9 +561,32 @@ def test_site_hourly_forecast_query_is_registered():
...
     }.issubset(query.placeholders)
+    assert (
+        "AVG(COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5)) AS pm25_mean"
+        in query.sql
+    )
+    assert "COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5) IS NOT NULL" in query.sql
     assert "HAVING COUNT(DISTINCT timestamp) >= {min_hours}" in query.sql
 
 
+def test_site_daily_forecast_query_falls_back_to_raw_pm25():
```
Comment (NicholasTurner23) [inline src/workflows/airqo_etl_utils/tests/test_site_hourly_forecast.py]: This is not a necessary test. @wabinyai

File: src/workflows/airqo_etl_utils/sql/forecasting/2026022002site_daily_aggregated.sql [hunk 2]
```diff
@@ -29,6 +29,7 @@ ORDER BY day, site_id;
 
 -- name: site_daily_aggregated_for_forecast_jobs
 -- Daily aggregated site PM2.5 for site-forecast prediction jobs with site metadata backfill.
+-- Uses calibrated PM2.5 when available, falling back to raw PM2.5.
```
Comment (NicholasTurner23) [inline src/workflows/airqo_etl_utils/sql/forecasting/2026022002site_daily_aggregated.sql:32]: Note that this will probably include non airqo device's data. In case you only want to use airqo data, you will have to filter by network on the "where". Refer to the schemas.

**Review by NicholasTurner23 — APPROVED** [review]
We can have this pushed to stage.

File: src/workflows/airqo_etl_utils/sql/forecasting/2026022002site_daily_aggregated.sql [hunk 3]
```diff
@@ -45,22 +46,23 @@ SELECT
...
+    MIN(COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5)) AS pm25_min,
+    MAX(COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5)) AS pm25_max,
     COUNT(DISTINCT TIMESTAMP_TRUNC(t1.timestamp, HOUR)) AS n_hours
 FROM {consolidated_table} AS t1
 LEFT JOIN {sites_table} AS t2
     ON t1.site_id = t2.id
 WHERE DATE(t1.timestamp) BETWEEN DATE('{start_date}') AND DATE('{end_date}')
     AND t1.site_id IS NOT NULL
-    AND t1.pm2_5_calibrated_value IS NOT NULL
+    AND COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5) IS NOT NULL
```
Comment (NicholasTurner23) [inline src/workflows/airqo_etl_utils/sql/forecasting/2026022002site_daily_aggregated.sql]: While this is semantically correct, I think it is better to use the expression where A is not null or B is not null.

File: src/workflows/airqo_etl_utils/ml_utils.py [hunk 4]
```diff
@@ -2941,6 +2947,79 @@ def with_empty_met_columns(frame: pd.DataFrame) -> pd.DataFrame:
 
         return with_empty_met_columns(enriched)
 
+    @staticmethod
```
Comment (NicholasTurner23) [inline src/workflows/airqo_etl_utils/ml_utils.py:2950]: Lets endeavor to improve on documentation. The one liner doc strings do not help much. @wabinyai

File: src/workflows/airqo_etl_utils/ml_utils.py [hunk 5]
```diff
@@ -2941,6 +2947,79 @@ def with_empty_met_columns(frame: pd.DataFrame) -> pd.DataFrame:
...
+
+        for key in key_columns:
+            filled[key] = pd.to_numeric(filled[key], errors="coerce")
+            met_lookup[key] = pd.to_numeric(met_lookup[key], errors="coerce")
+
+        row_order = "_forecast_row_order"
+        filled[row_order] = np.arange(len(filled))
+
+        nearest_frames: List[pd.DataFrame] = []
+        for coordinates, forecast_group in filled.groupby(key_columns, dropna=True):
```
Comment (NicholasTurner23) [inline src/workflows/airqo_etl_utils/ml_utils.py:2996]: We talked about nested loops. Not a good idea when dealing with lots of data @wabinyai

File: src/workflows/airqo_etl_utils/tests/test_site_hourly_forecast.py [hunk 6]
```diff
@@ -552,16 +640,71 @@ def test_met_no_hourly_payload_parser_includes_timestamp():
...
+    assert (
+        "MIN(COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5)) AS pm25_min"
+        in daily_query.sql
+    )
+    assert (
+        "MAX(COALESCE(t1.pm2_5_calibrated_value, t1.pm2_5)) AS pm25_max"
+        in daily_query.sql
+    )
+    assert "ANY_VALUE(t2.latitude)" in daily_query.sql
+    assert "ANY_VALUE(t2.approximate_latitude)" in daily_query.sql
```
Comment (NicholasTurner23) [inline src/workflows/airqo_etl_utils/tests/test_site_hourly_forecast.py:697]: Focus on testing functionality and edge cases vs testing string statements. @wabinyai

**Review by NicholasTurner23 — COMMENTED** [review]
I've left some comments for you @wabinyai

File: src/workflows/airqo_etl_utils/ml_utils.py (hunk 5 above)
Comment (wabinyai) [reply src/workflows/airqo_etl_utils/ml_utils.py:2996]: Currently, this works ok.
